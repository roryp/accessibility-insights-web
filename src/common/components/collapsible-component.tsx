// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@uifabric/utilities';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        const newState = !this.state.showContent;
        this.setState({ showContent: newState });
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, 'collapsible-content')}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName, 'collapsible-component')}>
                <ActionButton
                    className="collapsible"
                    iconProps={{ iconName: iconName, class: 'collapsible-icon' }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    <span className="collapsible-title">{this.props.header}</span>
                </ActionButton>
                {content}
            </div>
        );
    }
}
