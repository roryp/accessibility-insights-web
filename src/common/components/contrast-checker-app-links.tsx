// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { windowsProductName } from 'content/strings/application';
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { NewTabLink } from './new-tab-link';

export const WindowsContrastCheckerAppLink = NamedFC('WindowsContrastCheckerAppLink', () => (
    <NewTabLink href="https://go.microsoft.com/fwlink/?linkid=2075365">
        {windowsProductName}
    </NewTabLink>
));

export const MacContrastCheckerAppLink = NamedFC('MacContrastCheckerAppLink', () => (
    <NewTabLink href="https://developer.paciellogroup.com/resources/contrastanalyser/">
        Colour Contrast Analyser
    </NewTabLink>
));
