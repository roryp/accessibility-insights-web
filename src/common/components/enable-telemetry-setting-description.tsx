// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { PrivacyStatementText, PrivacyStatementTextDeps } from './privacy-statement-text';
import { TelemetryNotice, TelemetryNoticeDeps } from './telemetry-notice';

export type EnableTelemetrySettingDescriptionDeps = TelemetryNoticeDeps & PrivacyStatementTextDeps;

export type EnableTelemetrySettingDescriptionProps = {
    deps: EnableTelemetrySettingDescriptionDeps;
};

export const EnableTelemetrySettingDescription = NamedFC<EnableTelemetrySettingDescriptionProps>(
    'EnableTelemetrySettingDescription',
    props => (
        <>
            <TelemetryNotice deps={props.deps} />
            <PrivacyStatementText deps={props.deps} />
        </>
    ),
);
