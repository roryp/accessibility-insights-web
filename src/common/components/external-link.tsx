// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link } from 'office-ui-fabric-react';
import * as React from 'react';

import { ActionInitiators } from '../action/action-initiator';
import { NamedFC } from '../react/named-fc';

export type ExternalLinkDeps = {
    actionInitiators: Pick<ActionInitiators, 'openExternalLink'>;
};

export type ExternalLinkProps = {
    deps: ExternalLinkDeps;
    href: string;
    title?: string;
};

export const ExternalLink = NamedFC<ExternalLinkProps>(
    'ExternalLink',
    ({ deps, href, title, children }) => {
        const onClick = e => deps.actionInitiators.openExternalLink(e, { href });
        return (
            <Link
                className="insights-link"
                target="_blank"
                href={href}
                title={title}
                onClick={onClick}
            >
                {children}
            </Link>
        );
    },
);
