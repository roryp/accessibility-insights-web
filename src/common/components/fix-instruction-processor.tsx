// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import * as styles from './cards/fix-instruction-color-box.scss';

type ColorMatch = {
    splitIndex: number;
    colorHexValue: string;
};

export class FixInstructionProcessor {
    private readonly colorValueMatcher = `(#[0-9a-f]{6})`;
    private readonly foregroundColorText = 'foreground color: ';
    private readonly foregroundRegExp = new RegExp(
        `${this.foregroundColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundColorText = 'background color: ';
    private readonly backgroundRegExp = new RegExp(
        `${this.backgroundColorText}${this.colorValueMatcher}`,
        'i',
    );

    public process(fixInstruction: string): JSX.Element {
        const matches = this.getColorMatches(fixInstruction);
        return this.splitFixInstruction(fixInstruction, matches);
    }

    private getColorMatches(fixInstruction: string): ColorMatch[] {
        const foregroundMatch = this.getColorMatch(fixInstruction, this.foregroundRegExp);
        const backgroundMatch = this.getColorMatch(fixInstruction, this.backgroundRegExp);

        return [foregroundMatch, backgroundMatch].filter(match => match != null) as ColorMatch[];
    }

    private getColorMatch(fixInstruction: string, colorRegex: RegExp): ColorMatch | null {
        if (!colorRegex.test(fixInstruction)) {
            return null;
        }

        const match = colorRegex.exec(fixInstruction);
        if (match == null || match[1] == null) {
            return null;
        }

        const colorHexValue = match[1];

        return {
            splitIndex: match.index + this.foregroundColorText.length + colorHexValue.length,
            colorHexValue,
        };
    }

    private splitFixInstruction(fixInstruction: string, matches: ColorMatch[]): JSX.Element {
        const sortedMatches = matches.sort((a, b) => a.splitIndex - b.splitIndex);

        if (sortedMatches.length === 0) {
            return <>{fixInstruction}</>;
        }

        let insertionIndex = 0;
        let keyIndex = 0;

        const result: JSX.Element[] = [];

        sortedMatches.forEach(match => {
            const endIndex = match.splitIndex - match.colorHexValue.length;
            const substring = fixInstruction.substring(insertionIndex, endIndex);

            result.push(<span key={`instruction-split-${keyIndex++}`}>{substring}</span>);

            result.push(this.createColorBox(match.colorHexValue, keyIndex++));

            insertionIndex = endIndex;
        });

        const coda = fixInstruction.substr(insertionIndex);

        result.push(<span key={`instruction-split-${keyIndex++}`}>{coda}</span>);

        return (
            <>
                <span aria-hidden="true">{result}</span>
                <span className={styles.screenReaderOnly}>{fixInstruction}</span>
            </>
        );
    }

    private createColorBox(colorHexValue: string, keyIndex: number): JSX.Element {
        return (
            <span
                key={`instruction-split-${keyIndex}`}
                className={styles.fixInstructionColorBox}
                style={{ backgroundColor: colorHexValue }}
            />
        );
    }
}
