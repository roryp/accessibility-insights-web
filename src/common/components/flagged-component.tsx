// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { FeatureFlagStoreData } from '../types/store-data/feature-flag-store-data';

export interface FlaggedComponentProps {
    featureFlagStoreData: FeatureFlagStoreData;
    featureFlag: string;
    enableJSXElement: JSX.Element;
    disableJSXElement?: JSX.Element;
}

export class FlaggedComponent extends React.Component<FlaggedComponentProps> {
    public render(): JSX.Element | null {
        const flagName = this.props.featureFlag;

        if (this.props.featureFlagStoreData != null && this.props.featureFlagStoreData[flagName]) {
            return this.props.enableJSXElement;
        }

        return this.props.disableJSXElement || null;
    }
}
