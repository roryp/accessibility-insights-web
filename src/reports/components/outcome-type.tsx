// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { CheckIcon, CheckIconInverted } from 'common/icons/check-icon';
import { CircleIcon } from 'common/icons/circle-icon';
import { CrossIcon, CrossIconInverted } from 'common/icons/cross-icon';
import { InapplicableIcon, InapplicableIconInverted } from 'common/icons/inapplicable-icon';
import * as React from 'react';

import { InstanceOutcomeType } from './instance-outcome-type';
import { RequirementOutcomeType } from './requirement-outcome-type';

export type OutcomeUnits = 'percentage' | 'requirements';

export type OutcomeStats = { [OT in OutcomeType]: number };
export type OutcomeType = RequirementOutcomeType | InstanceOutcomeType;

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'Passed' },
    incomplete: { pastTense: 'Incomplete' },
    fail: { pastTense: 'Failed' },
    inapplicable: { pastTense: 'Not applicable' },
    review: { pastTense: 'Needs review' },
};

export const outcomeIconMap: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIcon />,
    incomplete: <CircleIcon />,
    fail: <CrossIcon />,
    inapplicable: <InapplicableIcon />,
    review: <CircleIcon />,
};

export const outcomeIconMapInverted: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIconInverted />,
    incomplete: <CircleIcon />,
    fail: <CrossIconInverted />,
    inapplicable: <InapplicableIconInverted />,
    review: <CircleIcon />,
};
