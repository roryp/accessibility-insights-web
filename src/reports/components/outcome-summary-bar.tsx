// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as classNames from 'classnames';
import { NamedFC } from 'common/react/named-fc';
import { kebabCase } from 'lodash';
import * as React from 'react';

import { outcomeIconMap, outcomeIconMapInverted, OutcomeStats, OutcomeType } from './outcome-type';

export type OutcomeSummaryBarProps = {
    outcomeStats: Partial<OutcomeStats>;
    allOutcomeTypes: OutcomeType[];
    iconStyleInverted?: boolean;
    countSuffix?: string;
};

export const OutcomeSummaryBar = NamedFC<OutcomeSummaryBarProps>('OutcomeSummaryBar', props => {
    const outcomeTypesCount = props.allOutcomeTypes.length;

    const getClassNames = (outcomeType: OutcomeType, index: number) => {
        return classNames({
            [kebabCase(outcomeType)]: true,
            'summary-bar-left-edge': index === 0,
            'summary-bar-right-edge': index === outcomeTypesCount - 1,
        });
    };

    return (
        <div className="outcome-summary-bar">
            {props.allOutcomeTypes.map((outcomeType, index) => {
                const { iconStyleInverted, countSuffix } = props;
                const iconMap =
                    iconStyleInverted === true ? outcomeIconMapInverted : outcomeIconMap;
                const outcomeIcon = iconMap[outcomeType];
                const count = props.outcomeStats[outcomeType];

                return (
                    <div key={outcomeType} style={{ flexGrow: count }}>
                        <span className={getClassNames(outcomeType, index)}>
                            <span aria-hidden="true">{outcomeIcon}</span>
                            {count}
                            {countSuffix}
                        </span>
                    </div>
                );
            })}
        </div>
    );
});
