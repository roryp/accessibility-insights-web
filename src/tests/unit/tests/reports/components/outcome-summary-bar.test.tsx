// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';
import { OutcomeSummaryBar, OutcomeSummaryBarProps } from 'reports/components/outcome-summary-bar';
import { OutcomeType } from 'reports/components/outcome-type';

describe('OutcomeSummaryBar', () => {
    const allOutcomeTypes: OutcomeType[] = ['pass', 'fail', 'incomplete'];
    const outcomeStats = {
        pass: 42,
        incomplete: 7,
        fail: 13,
    };

    it('show by percentage', () => {
        const props: OutcomeSummaryBarProps = { outcomeStats, allOutcomeTypes, countSuffix: '%' };
        const wrapper = shallow(<OutcomeSummaryBar {...props} />);

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('show by count', () => {
        const props: OutcomeSummaryBarProps = { outcomeStats, allOutcomeTypes };
        const wrapper = shallow(<OutcomeSummaryBar {...props} />);

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('render inverted badges', () => {
        const props: OutcomeSummaryBarProps = {
            outcomeStats,
            allOutcomeTypes,
            iconStyleInverted: true,
        };
        const wrapper = shallow(<OutcomeSummaryBar {...props} />);

        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
