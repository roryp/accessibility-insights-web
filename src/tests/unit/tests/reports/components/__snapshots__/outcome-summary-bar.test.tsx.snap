// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutcomeSummaryBar render inverted badges 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass summary-bar-left-edge"
    >
      <span
        aria-hidden="true"
      >
        <CheckIconInverted />
      </span>
      42
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIconInverted />
      </span>
      13
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete summary-bar-right-edge"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
    </span>
  </div>
</div>
`;

exports[`OutcomeSummaryBar show by count 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass summary-bar-left-edge"
    >
      <span
        aria-hidden="true"
      >
        <CheckIcon />
      </span>
      42
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIcon />
      </span>
      13
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete summary-bar-right-edge"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
    </span>
  </div>
</div>
`;

exports[`OutcomeSummaryBar show by percentage 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass summary-bar-left-edge"
    >
      <span
        aria-hidden="true"
      >
        <CheckIcon />
      </span>
      42
      %
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIcon />
      </span>
      13
      %
    </span>
  </div>
  <div
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete summary-bar-right-edge"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
      %
    </span>
  </div>
</div>
`;
